dask/dataframe/operation.py

```diff
@@ -413,7 +401,7 @@ def __init__(self, func, **kwargs):
         self.kwargs = kwargs
         self.cache = {}
 
-    def __call__(self, operation):
+    def __call__(self, operation, **kwargs):
```

Comment:
Careful here, `kwargs` are ignored in the cache key, so they aren't allowed to affect the result differently for two otherwise equal `operation`s.

If they need to, then taking a single hashable `arg` that the specific visitor can deconstruct works well IME:

```
def __call__(self, op, arg):
    key = (op, arg)
    ...

def _generate_graph(op, visitor, keys):
    dsk, dependency_keys = op.subgraph(dict(keys))
    for dep, dep_keys in dependency_keys.items():
        dsk.update(visitor(dep, keys=tuple(sorted(dep_keys.items()))
    return dsk

def generate_graph(op):
    return MemoizingVisitor(_generate_graph)(op, keys=tuple(sorted(op.collection_keys)))
```